From ec0b1c943070ef9d9533db3b569470586ad05c5a Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Mon, 20 Jan 2014 10:22:18 -0800
Subject: [PATCH 8/8] time: notify processes when time of day is changed

Enabling this will allow apps to register to be sent a signal
whenever the time of day on the system has changed (due to someone
calling settimeofday(), for instance).

This is exposed to userspace via /proc/notify_settime_signal

Signed-off-by: Peter Seebach <peter.seebach@windriver.com>
Integrated-by: Yong Zhang <yong.zhang@windriver.com>
Signed-off-by: Joe MacDonald <joe.macdonald@windriver.com>
[Adopt new proc fs APIs]
Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 fs/exec.c                 |    4 ++
 include/linux/init_task.h |    8 +++
 include/linux/sched.h     |    3 +
 include/linux/time.h      |    4 ++
 init/Kconfig              |   10 ++++
 kernel/Makefile           |    1 +
 kernel/hrtimer.c          |    4 ++
 kernel/notify_settime.c   |  113 +++++++++++++++++++++++++++++++++++++++++++++
 8 files changed, 147 insertions(+), 0 deletions(-)
 create mode 100644 kernel/notify_settime.c

diff --git a/fs/exec.c b/fs/exec.c
index 1f44670..9e913bd 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1549,6 +1549,10 @@ static int do_execve_common(const char *filename,
 	current->in_execve = 0;
 	acct_update_integrals(current);
 	free_bprm(bprm);
+#ifdef CONFIG_NOTIFY_SETTIME
+	/* there's got to be a better place for this!  */
+	current->settime_sig = 0;
+#endif
 	if (displaced)
 		put_files_struct(displaced);
 	return retval;
diff --git a/include/linux/init_task.h b/include/linux/init_task.h
index 5cd0f09..e4ea467 100644
--- a/include/linux/init_task.h
+++ b/include/linux/init_task.h
@@ -66,6 +66,13 @@ extern struct nsproxy init_nsproxy;
 	.signalfd_wqh	= __WAIT_QUEUE_HEAD_INITIALIZER(sighand.signalfd_wqh),	\
 }
 
+#ifdef CONFIG_NOTIFY_SETTIME
+#define INIT_NOTIFY_SETTIME \
+	.settime_sig    = 0,
+#else
+#define INIT_NOTIFY_SETTIME
+#endif
+
 extern struct group_info init_groups;
 
 #define INIT_STRUCT_PID {						\
@@ -222,6 +229,7 @@ extern struct task_group root_task_group;
 	INIT_TASK_RCU_PREEMPT(tsk)					\
 	INIT_CPUSET_SEQ							\
 	INIT_VTIME(tsk)							\
+	INIT_NOTIFY_SETTIME						\
 }
 
 
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 178a8d9..22e0293 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1416,6 +1416,9 @@ struct task_struct {
 	unsigned int	sequential_io;
 	unsigned int	sequential_io_avg;
 #endif
+#ifdef CONFIG_NOTIFY_SETTIME
+	int settime_sig;                /* notify_settime() signal */
+#endif
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
diff --git a/include/linux/time.h b/include/linux/time.h
index d5d229b..7903f42 100644
--- a/include/linux/time.h
+++ b/include/linux/time.h
@@ -181,6 +181,10 @@ extern s32 timekeeping_get_tai_offset(void);
 extern void timekeeping_set_tai_offset(s32 tai_offset);
 extern void timekeeping_clocktai(struct timespec *ts);
 
+#ifdef CONFIG_NOTIFY_SETTIME
+extern void do_notify_timechange(void);
+#endif
+
 struct tms;
 extern void do_sys_times(struct tms *);
 
diff --git a/init/Kconfig b/init/Kconfig
index 79addfc..41ee92f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1630,6 +1630,16 @@ config TRACEPOINTS
 
 source "arch/Kconfig"
 
+config NOTIFY_SETTIME
+	bool "Notify processes when time of day is changed"
+	default y
+	help
+	  Enabling this will allow apps to register to be sent a signal
+	  whenever the time of day on the system has changed (due to someone
+	  calling settimeofday(), for instance).
+
+	  This is exposed to userspace via /proc/notify_settime_signal
+
 endmenu		# General setup
 
 config HAVE_GENERIC_DMA_COHERENT
diff --git a/kernel/Makefile b/kernel/Makefile
index 5d7a07f..bf98627 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -103,6 +103,7 @@ obj-$(CONFIG_RING_BUFFER) += trace/
 obj-$(CONFIG_TRACEPOINTS) += trace/
 obj-$(CONFIG_IRQ_WORK) += irq_work.o
 obj-$(CONFIG_CPU_PM) += cpu_pm.o
+obj-$(CONFIG_NOTIFY_SETTIME) += notify_settime.o
 
 obj-$(CONFIG_PERF_EVENTS) += events/
 
diff --git a/kernel/hrtimer.c b/kernel/hrtimer.c
index 2288fbd..463f12f 100644
--- a/kernel/hrtimer.c
+++ b/kernel/hrtimer.c
@@ -771,6 +771,10 @@ void clock_was_set(void)
 	/* Retrigger the CPU local events everywhere */
 	on_each_cpu(retrigger_next_event, NULL, 1);
 #endif
+#ifdef CONFIG_NOTIFY_SETTIME
+	/* Notify any userspace apps that were interested in the change */
+	do_notify_timechange();
+#endif
 	timerfd_clock_was_set();
 }
 
diff --git a/kernel/notify_settime.c b/kernel/notify_settime.c
new file mode 100644
index 0000000..f516461
--- /dev/null
+++ b/kernel/notify_settime.c
@@ -0,0 +1,113 @@
+/*
+ *  Notify processes whenever the time is modified via settimeofday.
+ *  Copyright (C) 2006-2007 Wind River Systems, Inc.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <linux/sched.h>
+#include <linux/errno.h>
+#include <linux/signal.h>
+#include <linux/stat.h>
+#include <linux/proc_fs.h>
+#include <linux/seq_file.h>
+#include <linux/uaccess.h>
+
+/*
+ * Called when the time is modified below with settimeofday in order
+ * to notify all of the tasks that requested notification.
+ */
+void do_notify_timechange(void)
+{
+	struct task_struct *p;
+
+	read_lock(&tasklist_lock);
+	for_each_process(p) {
+		if (p->settime_sig)
+			group_send_sig_info(p->settime_sig, SEND_SIG_PRIV, p);
+	}
+	read_unlock(&tasklist_lock);
+}
+
+/*
+ * Set a signal to be delivered to the calling process whenever
+ * time of day is changed. A signal value of zero indicates that
+ * no signal should be sent.  On success, the call returns the
+ * old value of the signal.  A return value of -1 indicates an error.
+ */
+static int do_notify_settime(int sig)
+{
+	int ret = -EINVAL;
+
+	if ((sig > 0) && (sig <= _NSIG)) {
+		ret = current->settime_sig;
+		current->settime_sig = sig;
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_PROC_FS
+#define NOTIFY_SETTIME_PROCFS_NAME "notify_settime_signal"
+
+static int notify_settime_signal_show(struct seq_file *m, void *v )
+{
+	seq_printf(m, "%d\n", current->settime_sig); 
+	return 0;
+}
+
+static int notify_settime_signal_proc_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, notify_settime_signal_show, PDE_DATA(inode));
+}
+
+static int notify_settime_signal_proc_write(struct file *file,
+					const char __user *buffer,
+				       size_t count, loff_t *pos)
+{
+	char str[16];
+	int sig = 0;
+
+	if (count > sizeof(str)-1)
+		return -EINVAL;
+
+	memset(str, 0, sizeof(str));
+	if (copy_from_user(str, buffer, count))
+		return -EFAULT;
+
+	sscanf(str, "%d", &sig);
+	return do_notify_settime(sig);
+}
+
+static const struct file_operations notify_settime_signal_proc_fops = {
+	.open		= notify_settime_signal_proc_open,
+	.read		= seq_read,
+	.llseek		= seq_lseek,
+	.release	= single_release,
+	.write		= notify_settime_signal_proc_write,
+};
+ 
+static int __init notify_settime_init(void)
+{
+	int ret = 0;
+
+	proc_create(NOTIFY_SETTIME_PROCFS_NAME,
+		    S_IFREG | S_IRUGO | S_IWUSR, NULL,
+		    &notify_settime_signal_proc_fops);
+
+	return ret;
+}
+device_initcall(notify_settime_init);
+#endif
-- 
1.7.5.4

